inst/scribble/scribble.js

```diff
@@ -0,0 +1,626 @@
+    colorPicker.title = 'Choose Pen Color'
+    return colorPicker
+  }
+
+  createToolbox () {
+    this.toolBox = document.createElement('div')
+    this.toolBox.classList = 'xe-scribble__tools'
+  }
+
+  rgbColorToHex (color) {
+    // https://haacked.com/archive/2009/12/29/convert-rgb-to-hex.aspx/
+    if (color.substr(0, 1) === '#') {
+      return color
+    }
+    const digits = /rgb\((\d+), (\d+), (\d+)\)/.exec(color)
+
+    const red = parseInt(digits[1])
+    const green = parseInt(digits[2])
+    const blue = parseInt(digits[3])
+
+    const rgb = blue | (green << 8) | (red << 16)
+    return '#' + rgb.toString(16).padStart(6, '0')
+  }
+
+  pickContrastForegroundColor (color, light, dark, threshold) {
+    color = this.rgbColorToHex(color)
+
+    // https://stackoverflow.com/a/41491220/2022615
+    light = light || '#FFFFFF'
+    dark = dark || '#000000'
+    threshold = threshold || 145
+    color = (color.charAt(0) === '#') ? color.substring(1, 7) : color
+    const r = parseInt(color.substring(0, 2), 16) // hexToR
+    const g = parseInt(color.substring(2, 4), 16) // hexToG
+    const b = parseInt(color.substring(4, 6), 16) // hexToB
+    const score = ((r * 0.299) + (g * 0.587) + (b * 0.114))
+    return (score > threshold) ? dark : light
+  }
+
+  assembleToolbox () {
+    this.createToolbox()
+
+    // Build draw, erase, clear buttons
+    this.drawBtn = this.createButton('draw', 'draw', 'Start Drawing (S)')
+    this.eraseBtn = this.createButton('erase', 'eraser', 'Erase Lines')
+    this.clearBtn = this.createButton('clear', 'trash', 'Clear This Slide')
+
+    this.colorPicker = this.createColorPicker()
+    this.colorPicker.addEventListener('input', () => {
+      this.currColor = this.colorPicker.value
+      this.currFabric.freeDrawingBrush.color = this.currColor
+      this.setPenColorCSSVariables(this.currColor)
+    })
+
+    ;[this.drawBtn, this.eraseBtn, this.clearBtn, this.colorPicker]
+      .forEach((btn) => {
+        this.toolBox.appendChild(btn)
+      })
+
+    ;['click', 'touchend'].forEach((action) => {
+      this.toolBox.addEventListener(action, ev => ev.stopPropagation())
+    })
+  }
+
+  addToolboxToSlide () {
+    const canvasDiv = this.getVisibleSlideOuterContainer()
+    canvasDiv.appendChild(this.toolBox)
+    if (this.hideToolbox) {
+      this.toolBox.classList.add('minimized')
+    }
+  }
+
+  addToggleToolbox () {
+    const self = this
+
+    document.addEventListener('keydown', (ev) => {
+      if (ev.keyCode === this.launchKey) {
+        self.toggleToolbox()
+      }
+    })
+
+    this.toolBox.addEventListener('mouseleave', function (ev) {
+      if (self.drawMode | self.eraseMode) {
+        return
+      }
+      self.minimizeTimeout = setTimeout(() => self.toggleToolbox(false), 2000)
+    })
+
+    this.toolBox.addEventListener('mouseenter', function (ev) {
+      if (self.minimizeTimeout) {
+        clearTimeout(self.minimizeTimeout)
+      }
+    })
```

Comment:
How nicely does this play with touch devices? I'm not sure if you've tested this already, maybe this question isn't relevant.

Reply:
I just fake tested this by using Chrome devtools to use touch interactions and it works very well. The tools just wont minimize like they do for mouse interactions, which I think is fine.

Reply:
I tested on ipad/iphone and it works well. The toolbox does minimize, but you have to tap on the slide directly to clear the focus state of the last selected button.